Add common TLS configuration

backjo opened the pull request:
Signed-off-by: Jonah Back <jonah@jonahback.com>

Closes #1837 
<!--
We appreciate your contribution to the Jaeger project! 👋🎉

Before creating a pull request, please make sure:
- Your PR is solving one problem
- You have read the guide for contributing
  - See https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING.md
- You signed all your commits (otherwise we won't be able to merge the PR)
  - See https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md#certificate-of-origin---sign-your-work
- You added unit tests for the new functionality
- You mention in the PR description which issue it is addressing, e.g. "Resolves #123"
-->

## Which problem is this PR solving?
-  Makes TLS configuration standard across various storage systems

## Short description of the changes
- Moves TLS configurations into the tls package
- Refactors cassandra/kafka/elasticsearch backends to use the common configuration

backjo commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/jaegertracing/jaeger/1838)
<!-- Reviewable:end -->

yurishkuro commented:
@backjo Thanks!

I was also working on this on the plane, see #1840. It is slightly more configurable than your version, but so far I only switched gRPC client/server configs to it. Let me get it merged, and then you should be able to use that for storage configs.

NB We will need to add `SkipHostVerify` as another configurable option.

backjo commented:
Sounds good 👍

yurishkuro commented:
@backjo #1840 was merged, do you want to try to integrate it into this PR?

backjo commented:
@yurishkuro yeah, I'll try to integrate it in tonight

pavolloffay commented on plugin/storage/es/options_test.go:
```diff
@@ -57,7 +57,7 @@ func TestOptionsWithFlags(t *testing.T) {
 		"--es.aux.server-urls=3.3.3.3, 4.4.4.4",
 		"--es.aux.max-span-age=24h",
 		"--es.aux.num-replicas=10",
-		"--es.tls=true",
```
Breaking change! 

We should keep `--storage.tls=true`

objectiser replied:
+1 we need to keep the existing option, but can be marked as deprecated - and from now on use `es.tls.enabled` as the preferred way to enable TLS.

pavolloffay commented on pkg/config/tls/tls.go:
```diff
@@ -0,0 +1,131 @@
+
+const (
+	suffixTLS                  = ".tls"
+	suffixEnabled              = suffixTLS + ".enabled"
```
Breaking change see other comments.

pavolloffay commented:
The problem with current TLS options is that due to `--es.tls: true` and other embedded options, the config cannot be defined in JSON. In yaml there is a workaround: 

```yaml
es:
    tls: true
es.tls.cert: "/foo/bar"
```

So let's deprecate the `--es.tls` and if it is used use it but create a warning message. The tls package should not add deprecated enabled flag for new components (e.g tls for kafka)

yurishkuro commented on pkg/config/tls/tls.go:
```diff
@@ -0,0 +1,131 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tls
```
there is no need for this package, please use `pkg/config/tlscfg` from `master`. You can add `SkipHostVerify` option to `ClientFlagsConfig`

backjo commented:
Yeah, this isn't ready for review yet with your changes - let me update when it is

yurishkuro commented:
@backjo do you plan to continue this PR?

backjo commented:
yep - will try to push what I have locally tonight - sorry for the delay!

backjo commented:
@yurishkuro @pavolloffay I've updated this and I believe it's ready for review.

Summary of changes:
1) Move Kafka/ES/C* TLS setup and configuration to use the shared TLS config.
2) Added SkipHostVerify as a TLS option
3) Added a `tls.enabled: true` flag that is the recommended flag for enabling tls. `tls: true` is now marked as deprecated.

backjo commented:
@yurishkuro mind taking a look?

yurishkuro commented:
I'll take a look, most likely over the weekend, a bit swamped this week

yurishkuro commented on plugin/storage/cassandra/options.go:
```diff
@@ -81,9 +76,9 @@ func NewOptions(primaryNamespace string, otherNamespaces ...string) *Options {
-					EnableHostVerification: true,
+				TLS: tlscfg.Options{
+					Enabled:        false,
+					SkipHostVerify: false,
```
this is just setting zero values, can omit the TLS key completely

yurishkuro commented on plugin/storage/cassandra/options.go:
```diff
@@ -225,6 +203,11 @@ func (opt *Options) InitFromViper(v *viper.Viper) {
 }
 
 func (cfg *namespaceConfig) initFromViper(v *viper.Viper) {
+	var tlsFlagsConfig = tlscfg.ClientFlagsConfig{
```
please DRY it w/ L117, e.g. a helper function
```go
func tlsFlagsConfig(namespace string) tlscfg.ClientFlagsConfig
```

yurishkuro commented on plugin/storage/es/options.go:
```diff
@@ -112,6 +108,14 @@ func NewOptions(primaryNamespace string, otherNamespaces ...string) *Options {
 	return options
 }
 
+func (config *namespaceConfig) GetTLSFlagsConfig() *tlscfg.ClientFlagsConfig {
```
- does it need to be public?
- it can return config by value, we're not using pointers anywhere

yurishkuro commented on pkg/config/tlscfg/flags.go:
```diff
@@ -21,13 +21,16 @@ import (
-	tlsServerName  = "tls.server-name"
-	tlsClientCA    = "tls.client-ca"
-	tlsClientCAOld = "tls.client.ca"
+	tlsPrefix         = ".tls"
```
I am not seeing changes in the modules (grpc, etc) that are already using this package and don't assume that it adds a dot prefix. Seems like they will be broken by this change.

backjo commented:
Trying to tackle these review comments today - just getting back from a vacation

yurishkuro commented:
@backjo do you have the cycles to finish this PR? If not, I can take over.

backjo commented:
@yurishkuro will have cycles at the end of this week - otherwise, feel free to take over if you want to get it in sooner.

yurishkuro commented:
no hurry, if you can finish it it would be great.

pavolloffay commented:
A couple of remarks:

* cassandra and ES use different flags for host verification `--cassandra.tls.verify-host` and `--es.tls.skip-host-verify` - make sure both work. Pick one going forward and deprecate the other one
* GRPC reporter support `reporter.grpc.tls.server-name` which is not supported by other TLS impls at the moment.
* ES TLS impl support insecure (`--es.tls.skip-host-verify`) without any cert https://github.com/jaegertracing/jaeger/commit/47d20296982d2e01432be46daef1dfc2c365872b#diff-06fc7142fac7ac3ed9db42948f2558c8R306
* ES TLS impl support CA cert without any other certs and tls disabled https://github.com/jaegertracing/jaeger/commit/47d20296982d2e01432be46daef1dfc2c365872b#diff-06fc7142fac7ac3ed9db42948f2558c8R308

Make sure the functionality above works with the refactoring

backjo commented:
@pavolloffay I think we should go with 'skip-host-verify' as the default option , with 'verify-host' being the deprecated one, since the 'skip-host-verify' will be more secure with the default value.

backjo commented on pkg/es/config/config.go:
```diff
@@ -301,14 +291,6 @@ func (c *Configuration) getConfigOptions(logger *zap.Logger) ([]elastic.ClientOp
-				return nil, err
-			}
-			httpTransport.TLSClientConfig = &tls.Config{RootCAs: ca}
-		}
```
Yeah, you're correct. adding it back.

yurishkuro commented:
> I think we should go with 'skip-host-verify' as the default option , with 'verify-host' being the deprecated one, since the 'skip-host-verify' will be more secure with the default value.

+1

pavolloffay commented:
Some of the flags look weird - two dots: 
```
      --reporter.grpc..tls                                        (deprecated) see --reporter.grpc..tls.enabled
      --reporter.grpc..tls.ca string                              Path to a TLS CA (Certification Authority) file used to verify the remote server(s) (by default will use the system truststore)
      --reporter.grpc..tls.cert string                            Path to a TLS Certificate file, used to identify this process to the remote server(s)
      --reporter.grpc..tls.enabled                                Enable TLS when talking to the remote server(s)
      --reporter.grpc..tls.key string                             Path to a TLS Private Key file, used to identify this process to the remote server(s)
      --reporter.grpc..tls.server-name string                     Override the TLS server name we expect in the certificate of the remove server(s)
      --reporter.grpc..tls.skip-host-verify                       Skip TLS hostname verification
```

pavolloffay commented on pkg/es/config/config.go:
```diff
@@ -301,17 +291,14 @@ func (c *Configuration) getConfigOptions(logger *zap.Logger) ([]elastic.ClientOp
 		}
 	} else {
 		httpTransport := &http.Transport{
-			Proxy: http.ProxyFromEnvironment,
```
revert all the changes in this function except line 285. It breaks a couple of things  e.g. https://github.com/jaegertracing/jaeger/commit/47d20296982d2e01432be46daef1dfc2c365872b#diff-06fc7142fac7ac3ed9db42948f2558c8R306

backjo replied:
yep, i've got a TODO still to restore this

pavolloffay commented on pkg/config/tlscfg/options.go:
```diff
@@ -26,12 +26,13 @@ import (
+	KeyPath        string
+	ServerName     string // only for client-side TLS config
+	ClientCAPath   string // only for server-side TLS config for client auth
+	SkipHostVerify bool
```
On line 50 add `InsecureSkipVerify: p.SkipHostVerify,`

pavolloffay commented on pkg/es/config/config.go:
```diff
@@ -305,13 +296,12 @@ func (c *Configuration) getConfigOptions(logger *zap.Logger) ([]elastic.ClientOp
-			ctls := &TLSConfig{CaPath: c.TLS.CaPath}
-			ca, err := ctls.loadCertificate()
+		if c.TLS.CAPath != "" {
+			config, err := c.TLS.Config()
```
Here we want to load only CA cert. Does it load only CA. What happens is client cert or key are not specified?

yurishkuro commented on pkg/config/tlscfg/flags.go:
```diff
@@ -48,19 +51,22 @@ type ServerFlagsConfig struct {
 	if c.ShowServerName {
 		flags.String(c.Prefix+tlsServerName, "", "Override the TLS server name we expect in the certificate of the remove server(s)")
 	}
+	flags.Bool(c.Prefix+tlsSkipHostVerify, false, "Skip TLS hostname verification")
```
We had a better description before: `(insecure) Skip server's certificate chain and host name verification`

yurishkuro commented:
@backjo thanks for your patience, and sorry for the delay in reviewing. There are just a few small issues remaining:

(a) can you amend the commit b7d5be88a88e43e3aa0728147b6965d228b8f98c to include the sign-off?

(b) These flags are missing, need to restore and deprecate them:
```
       --cassandra-archive.tls.verify-host
       --cassandra.tls.verify-host
```

(c) sync with master

backjo commented:
@yurishkuro will do - I think I'm going to rebase the changes into one commit and bring it current with master. 

Re: verify-host - I'm a bit uncertain how we're going to handle conflicting behavior between the two, since the flags are inverted. Specifically, thinking about these cases:

1) User does not set  `--cassandra.tls.verify-host`, which implicitly sets it to false.  This default behavior conflicts with the default behavior of skip-host-verification, which would be to implicitly enable host verification. 
2) User explicitly sets `--cassandra.tls.verify-host` to false. This could be backward-compatible / respected by using the `IsSet` call in Viper
3) User explicitly sets `--cassandra.tls.verify-host` to true. This aligns with the default value of skip-host-verification so the two would be in alignment. If both were explicitly set for some reason, we would respect skip-host-verification. 

The first case is a bit tricky - the most backward-compatible way would be to keep the host verification off by default - which would mean that users would need to explicitly enable hostname verification with the new property, until some time later on when we decide that we should enable the more secure option by default.

yurishkuro commented:
I think it is ok to change the behavior to verify bu default. We will document it in the release notes as a breaking change. The legacy option is needed to that people's start scripts don't break (they won't be affected by the default behavior change since they're already specifying the legacy flag explicitly).

yurishkuro commented:
@backjo one last push to get this over the line?

yurishkuro commented on Gopkg.toml:
```diff
@@ -97,7 +97,7 @@ required = [
 [[constraint]]
   name = "github.com/spf13/viper"
-  version = "^1.0.0"
+  version = "^1.5.0"
```
Is this change necessary? If we're changing this file, we need to run `dep ensure --update` and also check-in the lock file.

backjo replied:
It is - there was a bug in Viper where `isSet` was returning true even in cases where it was not set, but there was a default value. This was fixed in 1.5

yurishkuro commented:
@pavolloffay are you ok to approve/merge?

pavolloffay commented:
I have just tested the following ES configurations against elastic.co operator. So far all works as before.

to test skip-host-verify
```
cat <<EOF | kubectl apply -f -
apiVersion: jaegertracing.io/v1
kind: Jaeger
metadata:
  name: simple-prod
spec:
  strategy: production
  collector:
    image: pavolloffay/jaeger-collector:es-tls2
  query:
    image: pavolloffay/jaeger-query:es-tls2
  storage:
    type: elasticsearch
    options:
      es:
        server-urls: https://quickstart-es-http:9200
        tls:
          skip-host-verify: true
        num-shards: 1
        num-replicas: 0
    secretName: jaeger-secret
  volumeMounts:
    - name: certificates
      mountPath: /es/certificates/
      readOnly: true
  volumes:
    - name: certificates
      secret:
        secretName: quickstart-es-http-certs-public
EOF
``` 

To test CA cert without providing all certs.
```
cat <<EOF | kubectl apply -f -
apiVersion: jaegertracing.io/v1
kind: Jaeger
metadata:
  name: simple-prod
spec:
  strategy: production
  collector:
    image: pavolloffay/jaeger-collector:es-tls2
  query:
    image: pavolloffay/jaeger-query:es-tls2
  storage:
    type: elasticsearch
    options:
      es:
        server-urls: https://quickstart-es-http:9200
        tls:
          skip-host-verify: true
        num-shards: 1
        num-replicas: 0
    secretName: jaeger-secret
  volumeMounts:
    - name: certificates
      mountPath: /es/certificates/
      readOnly: true
  volumes:
    - name: certificates
      secret:
        secretName: quickstart-es-http-certs-public
EOF
```

pavolloffay commented on plugin/storage/es/options.go:
```diff
@@ -112,6 +108,14 @@ func NewOptions(primaryNamespace string, otherNamespaces ...string) *Options {
+	return tlscfg.ClientFlagsConfig{
+		Prefix:         config.namespace,
+		ShowEnabled:    true,
+		ShowServerName: true,
```
This flag wasn't supported before.

yurishkuro commented:
Thank you, @backjo!

jpkrohling commented on pkg/kafka/auth/tls.go:
```diff
@@ -15,57 +15,20 @@
-	if err != nil {
-		return errors.Wrap(err, "error loading tls config")
+func setTLSConfiguration(config *tlscfg.Options, saramaConfig *sarama.Config) error {
+	if config.Enabled {
```
This is a breaking change: previously, there was no `--kafka.consumer.tls` option (or `producer`), and TLS config would be used on whether the `getTLS()` function returned an error or not. Now, even if all the parameters are set, the collector/ingester won't be using TLS unless a new option is added to the command line, `--kafka.consumer.tls.enabled` (or the one marked as deprecated but that never existed: `--kafka.consumer.tls`).

pavolloffay replied:
What is the "new" set of parameters to make it work?

jpkrohling replied:
The `tls.enabled` ones: `kafka.consumer.tls.enabled` and `kafka.producer.tls.enabled`.

objectiser replied:
So to be backward compatible, it would have to use the `getTLS()` approach still - possibly with the `getTLS()` method checking the enabled flag (i.e. if enabled = false then don't use)?

jpkrohling replied:
That would work, unless "not specified" means "false" when checking the `tls.enabled` flag.

objectiser replied:
Yep - unfortunately `tls.enabled` default is false.

jpkrohling replied:
Created #2092 to track this.
